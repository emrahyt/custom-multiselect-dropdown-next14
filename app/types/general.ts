export interface SimplifiedCharacter {
  name: string;
  id: number;
  image: string;
  episode: string[];
}

export type ParsedCharacter = Omit<SimplifiedCharacter, "episode"> & {
  numberOfEpisodes: number;
};

export type CharacterBaseInfo = Pick<ParsedCharacter, "id" | "name">;

export type MultiselectCharacter = Pick<
  ParsedCharacter,
  "id" | "name" | "image"
>;

export interface ParsedCharactersResponse {
  totalPages: number;
  charactersList: ParsedCharacter[];
}
