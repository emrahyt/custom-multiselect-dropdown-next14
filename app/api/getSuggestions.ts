import axios, { AxiosResponse, AxiosError } from "axios";
import { BASE_URL, INTERNAL_SERVER_ERROR } from "utils/constants";
import { parseCharactersResponse } from "utils/helpers";
import { ParsedCharactersResponse } from "@/types/general";
import { ApiResponse, ApiError } from "@/types/api";

interface CustomResponse<T> {
  data?: T;
  status?: number;
  error?: string;
}

export const getSuggestions = async (
  characterName: string,
  page: number
): Promise<CustomResponse<ParsedCharactersResponse>> => {
  try {
    const response: AxiosResponse<ApiResponse> = await axios.get(
      `${BASE_URL}/character/?page=${page}&name=${characterName}`
    );

    const parsedResponse = parseCharactersResponse(response.data);

    return { data: parsedResponse, status: response.status };
  } catch (error) {
    const axiosError = error as AxiosError<ApiError>;

    const customResponse: CustomResponse<ParsedCharactersResponse> = {
      data: undefined,
      status: axiosError.response?.status || 500,
      error: axiosError.response?.data?.error || INTERNAL_SERVER_ERROR,
    };

    return customResponse;
  }
};
