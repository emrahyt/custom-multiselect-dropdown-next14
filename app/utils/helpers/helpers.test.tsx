import { ParsedCharactersResponse, CharacterBaseInfo } from "@/types/general";
import { ApiResponse } from "@/types/api";
import {
  parseCharactersResponse,
  highlightText,
  handleFormerSelections,
} from ".";

describe("parseCharactersResponse", () => {
  it("should parse characters response correctly", () => {
    const mockCharactersResponse: ApiResponse = {
      info: {
        count: 10,
        pages: 2,
        next: "next-page-url",
        prev: "prev-page-url",
      },
      results: [
        {
          id: 1,
          name: "Rick",
          status: "Alive",
          species: "Human",
          type: "Main",
          gender: "Male",
          origin: { name: "Earth", url: "earth-url" },
          location: { name: "Location 1", url: "location1-url" },
          image: "image-url",
          episode: ["episode1", "episode2"],
          url: "character1-url",
          created: "2024-01-20",
        },
      ],
    };

    const expectedParsedResponse: ParsedCharactersResponse = {
      totalPages: 2,
      charactersList: [
        {
          id: 1,
          name: "Rick",
          image: "image-url",
          numberOfEpisodes: 2,
        },
      ],
    };

    const actualParsedResponse = parseCharactersResponse(
      mockCharactersResponse
    );

    expect(actualParsedResponse).toEqual(expectedParsedResponse);
  });
});

describe("highlightText", () => {
  it("should highlight search term in text", () => {
    const text = "This is a sample text for testing.";
    const searchTerm = "sample";
    const actualOutput = highlightText(text, searchTerm);
    expect(actualOutput).toMatchSnapshot();
  });

  it("should return original text if search term is empty", () => {
    const text = "This is a sample text for testing.";
    const searchTerm = "";
    const expectedOutput = text;
    const actualOutput = highlightText(text, searchTerm);
    expect(actualOutput).toEqual(expectedOutput);
  });

  it("should handle case-insensitive search", () => {
    const text = "This is a sample text for testing.";
    const searchTerm = "SAMPLE";
    const actualOutput = highlightText(text, searchTerm);
    expect(actualOutput).toMatchSnapshot();
  });
});

describe("handleFormerSelections", () => {
  it("should filter selected characters based on the available characters", () => {
    const selectedCharacters = [
      { id: 1, name: "Character 1" },
      { id: 2, name: "Character 2" },
      { id: 3, name: "Character 3" },
    ];

    const characters = [
      { id: 1, name: "Character 1", numberOfEpisodes: 3, image: "path1" },
      { id: 4, name: "Character 4", numberOfEpisodes: 1, image: "path2" },
      { id: 3, name: "Character 3", numberOfEpisodes: 5, image: "path3" },
    ];

    const expectedFormerSelections = [
      { id: 1, name: "Character 1" },
      { id: 3, name: "Character 3" },
    ];

    const actualFormerSelections = handleFormerSelections({
      selectedCharacters,
      characters,
    });

    expect(actualFormerSelections).toEqual(expectedFormerSelections);
  });

  it("should return an empty array if no characters are selected", () => {
    const selectedCharacters: CharacterBaseInfo[] = [];
    const characters = [
      { id: 1, name: "Character 1", numberOfEpisodes: 3, image: "path1" },
      { id: 4, name: "Character 4", numberOfEpisodes: 1, image: "path2" },
      { id: 3, name: "Character 3", numberOfEpisodes: 5, image: "path3" },
    ];

    const expectedFormerSelections: CharacterBaseInfo[] = [];
    const actualFormerSelections = handleFormerSelections({
      selectedCharacters,
      characters,
    });

    expect(actualFormerSelections).toEqual(expectedFormerSelections);
  });

  it("should return an empty array if no characters match the selected characters", () => {
    const selectedCharacters = [
      { id: 5, name: "Character 5" },
      { id: 6, name: "Character 6" },
    ];

    const characters = [
      { id: 1, name: "Character 1", numberOfEpisodes: 3, image: "path1" },
      { id: 4, name: "Character 4", numberOfEpisodes: 1, image: "path2" },
      { id: 3, name: "Character 3", numberOfEpisodes: 5, image: "path3" },
    ];

    const expectedFormerSelections: CharacterBaseInfo[] = [];

    const actualFormerSelections = handleFormerSelections({
      selectedCharacters,
      characters,
    });

    expect(actualFormerSelections).toEqual(expectedFormerSelections);
  });
});
