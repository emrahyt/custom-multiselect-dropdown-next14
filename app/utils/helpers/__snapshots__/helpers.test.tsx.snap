// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`highlightText should handle case-insensitive search 1`] = `
[
  "This is a ",
  <strong>
    sample
  </strong>,
  " text for testing.",
]
`;

exports[`highlightText should highlight search term in text 1`] = `
[
  "This is a ",
  <strong>
    sample
  </strong>,
  " text for testing.",
]
`;
